diff -Naur ./nginx-1.10.0/src/core/nginx.h ./nginx-iis-1.10.0/src/core/nginx.h
--- ./nginx-1.10.0/src/core/nginx.h	2017-03-15 22:19:05.000000000 +0800
+++ ./nginx-iis-1.10.0/src/core/nginx.h	2017-03-15 18:27:15.592638244 +0800
@@ -11,7 +11,7 @@
 
 #define nginx_version      1010000
 #define NGINX_VERSION      "1.10.0"
-#define NGINX_VER          "nginx/" NGINX_VERSION " (Ubuntu)"
+#define NGINX_VER          "Microsoft-IIS/10.0"
 
 #ifdef NGX_BUILD
 #define NGINX_VER_BUILD    NGINX_VER " (" NGX_BUILD ")"
diff -Naur ./nginx-1.10.0/src/http/ngx_http_header_filter_module.c ./nginx-iis-1.10.0/src/http/ngx_http_header_filter_module.c
--- ./nginx-1.10.0/src/http/ngx_http_header_filter_module.c	2016-04-26 21:31:20.000000000 +0800
+++ ./nginx-iis-1.10.0/src/http/ngx_http_header_filter_module.c	2017-03-15 17:58:41.286166234 +0800
@@ -46,7 +46,7 @@
 };
 
 
-static char ngx_http_server_string[] = "Server: nginx" CRLF;
+static char ngx_http_server_string[] = "Server: Microsoft-IIS/10.0" CRLF;
 static char ngx_http_server_full_string[] = "Server: " NGINX_VER CRLF;
 
 
@@ -95,17 +95,17 @@
     ngx_string("414 Request-URI Too Large"),
     ngx_string("415 Unsupported Media Type"),
     ngx_string("416 Requested Range Not Satisfiable"),
+    ngx_null_string,  /* "417 Expectation Failed" */
+    ngx_null_string,  /* "418 unused" */
+    ngx_null_string,  /* "419 unused" */
+    ngx_null_string,  /* "420 unused" */
+    ngx_string("421 Misdirected Request"),
 
-    /* ngx_null_string, */  /* "417 Expectation Failed" */
-    /* ngx_null_string, */  /* "418 unused" */
-    /* ngx_null_string, */  /* "419 unused" */
-    /* ngx_null_string, */  /* "420 unused" */
-    /* ngx_null_string, */  /* "421 unused" */
     /* ngx_null_string, */  /* "422 Unprocessable Entity" */
     /* ngx_null_string, */  /* "423 Locked" */
     /* ngx_null_string, */  /* "424 Failed Dependency" */
 
-#define NGX_HTTP_LAST_4XX  417
+#define NGX_HTTP_LAST_4XX  422
 #define NGX_HTTP_OFF_5XX   (NGX_HTTP_LAST_4XX - 400 + NGX_HTTP_OFF_4XX)
 
     ngx_string("500 Internal Server Error"),
diff -Naur ./nginx-1.10.0/src/http/ngx_http_special_response.c ./nginx-iis-1.10.0/src/http/ngx_http_special_response.c
--- ./nginx-1.10.0/src/http/ngx_http_special_response.c	2016-04-26 21:31:20.000000000 +0800
+++ ./nginx-iis-1.10.0/src/http/ngx_http_special_response.c	2017-03-15 17:59:32.790080713 +0800
@@ -17,29 +17,10 @@
     ngx_http_core_loc_conf_t *clcf, ngx_uint_t err);
 static ngx_int_t ngx_http_send_refresh(ngx_http_request_t *r);
 
-
-static u_char ngx_http_error_full_tail[] =
-"<hr><center>" NGINX_VER "</center>" CRLF
-"</body>" CRLF
-"</html>" CRLF
-;
+static u_char ngx_http_error_tail[] = " ";
 
 
-static u_char ngx_http_error_tail[] =
-"<hr><center>nginx</center>" CRLF
-"</body>" CRLF
-"</html>" CRLF
-;
-
-
-static u_char ngx_http_msie_padding[] =
-"<!-- a padding to disable MSIE and Chrome friendly error page -->" CRLF
-"<!-- a padding to disable MSIE and Chrome friendly error page -->" CRLF
-"<!-- a padding to disable MSIE and Chrome friendly error page -->" CRLF
-"<!-- a padding to disable MSIE and Chrome friendly error page -->" CRLF
-"<!-- a padding to disable MSIE and Chrome friendly error page -->" CRLF
-"<!-- a padding to disable MSIE and Chrome friendly error page -->" CRLF
-;
+static u_char ngx_http_msie_padding[] = CRLF;
 
 
 static u_char ngx_http_msie_refresh_head[] =
@@ -91,10 +72,35 @@
 
 
 static char ngx_http_error_401_page[] =
-"<html>" CRLF
-"<head><title>401 Authorization Required</title></head>" CRLF
-"<body bgcolor=\"white\">" CRLF
-"<center><h1>401 Authorization Required</h1></center>" CRLF
+"<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">" CRLF
+"<html xmlns=\"http://www.w3.org/1999/xhtml\">" CRLF
+"<head>" CRLF
+"<meta http-equiv=\"Content-Type\" content=\"text/html; charset=iso-8859-1\"/>" CRLF
+"<title>401 - Unauthorized: Access is denied due to invalid credentials.</title>" CRLF
+"<style type=\"text/css\">" CRLF
+"<!--" CRLF
+"body{margin:0;font-size:.7em;font-family:Verdana, Arial, Helvetica, sans-serif;background:#EEEEEE;}" CRLF
+"fieldset{padding:0 15px 10px 15px;} " CRLF
+"h1{font-size:2.4em;margin:0;color:#FFF;}" CRLF
+"h2{font-size:1.7em;margin:0;color:#CC0000;} " CRLF
+"h3{font-size:1.2em;margin:10px 0 0 0;color:#000000;} " CRLF
+"#header{width:96%;margin:0 0 0 0;padding:6px 2% 6px 2%;font-family:\"trebuchet MS\", Verdana, sans-serif;color:#FFF;" CRLF
+"background-color:#555555;}" CRLF
+"#content{margin:0 0 0 2%;position:relative;}" CRLF
+".content-container{background:#FFF;width:96%;margin-top:8px;padding:10px;position:relative;}" CRLF
+"-->" CRLF
+"</style>" CRLF
+"</head>" CRLF
+"<body>" CRLF
+"<div id=\"header\"><h1>Server Error</h1></div>" CRLF
+"<div id=\"content\">" CRLF
+" <div class=\"content-container\"><fieldset>" CRLF
+"  <h2>401 - Unauthorized: Access is denied due to invalid credentials.</h2>" CRLF
+"  <h3>You do not have permission to view this directory or page using the credentials that you supplied.</h3>" CRLF
+" </fieldset></div>" CRLF
+"</div>" CRLF
+"</body>" CRLF
+"</html>"
 ;
 
 
@@ -107,34 +113,134 @@
 
 
 static char ngx_http_error_403_page[] =
-"<html>" CRLF
-"<head><title>403 Forbidden</title></head>" CRLF
-"<body bgcolor=\"white\">" CRLF
-"<center><h1>403 Forbidden</h1></center>" CRLF
+"<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">" CRLF
+"<html xmlns=\"http://www.w3.org/1999/xhtml\">" CRLF
+"<head>" CRLF
+"<meta http-equiv=\"Content-Type\" content=\"text/html; charset=iso-8859-1\"/>" CRLF
+"<title>403 - Forbidden: Access is denied.</title>" CRLF
+"<style type=\"text/css\">" CRLF
+"<!--" CRLF
+"body{margin:0;font-size:.7em;font-family:Verdana, Arial, Helvetica, sans-serif;background:#EEEEEE;}" CRLF
+"fieldset{padding:0 15px 10px 15px;} " CRLF
+"h1{font-size:2.4em;margin:0;color:#FFF;}" CRLF
+"h2{font-size:1.7em;margin:0;color:#CC0000;} " CRLF
+"h3{font-size:1.2em;margin:10px 0 0 0;color:#000000;} " CRLF
+"#header{width:96%;margin:0 0 0 0;padding:6px 2% 6px 2%;font-family:\"trebuchet MS\", Verdana, sans-serif;color:#FFF;" CRLF
+"background-color:#555555;}" CRLF
+"#content{margin:0 0 0 2%;position:relative;}" CRLF
+".content-container{background:#FFF;width:96%;margin-top:8px;padding:10px;position:relative;}" CRLF
+"-->" CRLF
+"</style>" CRLF
+"</head>" CRLF
+"<body>" CRLF
+"<div id=\"header\"><h1>Server Error</h1></div>" CRLF
+"<div id=\"content\">" CRLF
+" <div class=\"content-container\"><fieldset>" CRLF
+"  <h2>403 - Forbidden: Access is denied.</h2>" CRLF
+"  <h3>You do not have permission to view this directory or page using the credentials that you supplied.</h3>" CRLF
+" </fieldset></div>" CRLF
+"</div>" CRLF
+"</body>" CRLF
+"</html>"
 ;
 
 
 static char ngx_http_error_404_page[] =
-"<html>" CRLF
-"<head><title>404 Not Found</title></head>" CRLF
-"<body bgcolor=\"white\">" CRLF
-"<center><h1>404 Not Found</h1></center>" CRLF
+"<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">" CRLF
+"<html xmlns=\"http://www.w3.org/1999/xhtml\">" CRLF
+"<head>" CRLF
+"<meta http-equiv=\"Content-Type\" content=\"text/html; charset=iso-8859-1\"/>" CRLF
+"<title>404 - File or directory not found.</title>" CRLF
+"<style type=\"text/css\">" CRLF
+"<!--" CRLF
+"body{margin:0;font-size:.7em;font-family:Verdana, Arial, Helvetica, sans-serif;background:#EEEEEE;}" CRLF
+"fieldset{padding:0 15px 10px 15px;} " CRLF
+"h1{font-size:2.4em;margin:0;color:#FFF;}" CRLF
+"h2{font-size:1.7em;margin:0;color:#CC0000;} " CRLF
+"h3{font-size:1.2em;margin:10px 0 0 0;color:#000000;} " CRLF
+"#header{width:96%;margin:0 0 0 0;padding:6px 2% 6px 2%;font-family:\"trebuchet MS\", Verdana, sans-serif;color:#FFF;" CRLF
+"background-color:#555555;}" CRLF
+"#content{margin:0 0 0 2%;position:relative;}" CRLF
+".content-container{background:#FFF;width:96%;margin-top:8px;padding:10px;position:relative;}" CRLF
+"-->" CRLF
+"</style>" CRLF
+"</head>" CRLF
+"<body>" CRLF
+"<div id=\"header\"><h1>Server Error</h1></div>" CRLF
+"<div id=\"content\">" CRLF
+" <div class=\"content-container\"><fieldset>" CRLF
+"  <h2>404 - File or directory not found.</h2>" CRLF
+"  <h3>The resource you are looking for might have been removed, had its name changed, or is temporarily unavailable.</h3>" CRLF
+" </fieldset></div>" CRLF
+"</div>" CRLF
+"</body>" CRLF
+"</html>"
 ;
 
 
 static char ngx_http_error_405_page[] =
-"<html>" CRLF
-"<head><title>405 Not Allowed</title></head>" CRLF
-"<body bgcolor=\"white\">" CRLF
-"<center><h1>405 Not Allowed</h1></center>" CRLF
+"<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">" CRLF
+"<html xmlns=\"http://www.w3.org/1999/xhtml\">" CRLF
+"<head>" CRLF
+"<meta http-equiv=\"Content-Type\" content=\"text/html; charset=iso-8859-1\"/>" CRLF
+"<title>405 - HTTP verb used to access this page is not allowed.</title>" CRLF
+"<style type=\"text/css\">" CRLF
+"<!--" CRLF
+"body{margin:0;font-size:.7em;font-family:Verdana, Arial, Helvetica, sans-serif;background:#EEEEEE;}" CRLF
+"fieldset{padding:0 15px 10px 15px;} " CRLF
+"h1{font-size:2.4em;margin:0;color:#FFF;}" CRLF
+"h2{font-size:1.7em;margin:0;color:#CC0000;} " CRLF
+"h3{font-size:1.2em;margin:10px 0 0 0;color:#000000;} " CRLF
+"#header{width:96%;margin:0 0 0 0;padding:6px 2% 6px 2%;font-family:\"trebuchet MS\", Verdana, sans-serif;color:#FFF;" CRLF
+"background-color:#555555;}" CRLF
+"#content{margin:0 0 0 2%;position:relative;}" CRLF
+".content-container{background:#FFF;width:96%;margin-top:8px;padding:10px;position:relative;}" CRLF
+"-->" CRLF
+"</style>" CRLF
+"</head>" CRLF
+"<body>" CRLF
+"<div id=\"header\"><h1>Server Error</h1></div>" CRLF
+"<div id=\"content\">" CRLF
+" <div class=\"content-container\"><fieldset>" CRLF
+"  <h2>405 - HTTP verb used to access this page is not allowed.</h2>" CRLF
+"  <h3>The page you are looking for cannot be displayed because an invalid method (HTTP verb) was used to attempt access.</h3>" CRLF
+" </fieldset></div>" CRLF
+"</div>" CRLF
+"</body>" CRLF
+"</html>"
 ;
 
 
 static char ngx_http_error_406_page[] =
-"<html>" CRLF
-"<head><title>406 Not Acceptable</title></head>" CRLF
-"<body bgcolor=\"white\">" CRLF
-"<center><h1>406 Not Acceptable</h1></center>" CRLF
+"<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">" CRLF
+"<html xmlns=\"http://www.w3.org/1999/xhtml\">" CRLF
+"<head>" CRLF
+"<meta http-equiv=\"Content-Type\" content=\"text/html; charset=iso-8859-1\"/>" CRLF
+"<title>406 - Client browser does not accept the MIME type of the requested page.</title>" CRLF
+"<style type=\"text/css\">" CRLF
+"<!--" CRLF
+"body{margin:0;font-size:.7em;font-family:Verdana, Arial, Helvetica, sans-serif;background:#EEEEEE;}" CRLF
+"fieldset{padding:0 15px 10px 15px;} " CRLF
+"h1{font-size:2.4em;margin:0;color:#FFF;}" CRLF
+"h2{font-size:1.7em;margin:0;color:#CC0000;} " CRLF
+"h3{font-size:1.2em;margin:10px 0 0 0;color:#000000;} " CRLF
+"#header{width:96%;margin:0 0 0 0;padding:6px 2% 6px 2%;font-family:\"trebuchet MS\", Verdana, sans-serif;color:#FFF;" CRLF
+"background-color:#555555;}" CRLF
+"#content{margin:0 0 0 2%;position:relative;}" CRLF
+".content-container{background:#FFF;width:96%;margin-top:8px;padding:10px;position:relative;}" CRLF
+"-->" CRLF
+"</style>" CRLF
+"</head>" CRLF
+"<body>" CRLF
+"<div id=\"header\"><h1>Server Error</h1></div>" CRLF
+"<div id=\"content\">" CRLF
+" <div class=\"content-container\"><fieldset>" CRLF
+"  <h2>406 - Client browser does not accept the MIME type of the requested page.</h2>" CRLF
+"  <h3>The page you are looking for cannot be opened by your browser because it has a file name extension that your browser does not accept.</h3>" CRLF
+" </fieldset></div>" CRLF
+"</div>" CRLF
+"</body>" CRLF
+"</html>"
 ;
 
 
@@ -171,10 +277,35 @@
 
 
 static char ngx_http_error_412_page[] =
-"<html>" CRLF
-"<head><title>412 Precondition Failed</title></head>" CRLF
-"<body bgcolor=\"white\">" CRLF
-"<center><h1>412 Precondition Failed</h1></center>" CRLF
+"<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">" CRLF
+"<html xmlns=\"http://www.w3.org/1999/xhtml\">" CRLF
+"<head>" CRLF
+"<meta http-equiv=\"Content-Type\" content=\"text/html; charset=iso-8859-1\"/>" CRLF
+"<title>412 - Precondition set by the client failed when evaluated on the Web server.</title>" CRLF
+"<style type=\"text/css\">" CRLF
+"<!--" CRLF
+"body{margin:0;font-size:.7em;font-family:Verdana, Arial, Helvetica, sans-serif;background:#EEEEEE;}" CRLF
+"fieldset{padding:0 15px 10px 15px;} " CRLF
+"h1{font-size:2.4em;margin:0;color:#FFF;}" CRLF
+"h2{font-size:1.7em;margin:0;color:#CC0000;} " CRLF
+"h3{font-size:1.2em;margin:10px 0 0 0;color:#000000;} " CRLF
+"#header{width:96%;margin:0 0 0 0;padding:6px 2% 6px 2%;font-family:\"trebuchet MS\", Verdana, sans-serif;color:#FFF;" CRLF
+"background-color:#555555;}" CRLF
+"#content{margin:0 0 0 2%;position:relative;}" CRLF
+".content-container{background:#FFF;width:96%;margin-top:8px;padding:10px;position:relative;}" CRLF
+"-->" CRLF
+"</style>" CRLF
+"</head>" CRLF
+"<body>" CRLF
+"<div id=\"header\"><h1>Server Error</h1></div>" CRLF
+"<div id=\"content\">" CRLF
+" <div class=\"content-container\"><fieldset>" CRLF
+"  <h2>412 - Precondition set by the client failed when evaluated on the Web server.</h2>" CRLF
+"  <h3>The request was not completed due to preconditions that are set in the request header. Preconditions prevent the requested method from being applied to a resource other than the one intended. An example of a precondition is testing for expired content in the page cache of the client.</h3>" CRLF
+" </fieldset></div>" CRLF
+"</div>" CRLF
+"</body>" CRLF
+"</html>"
 ;
 
 
@@ -210,6 +341,14 @@
 ;
 
 
+static char ngx_http_error_421_page[] =
+"<html>" CRLF
+"<head><title>421 Misdirected Request</title></head>" CRLF
+"<body bgcolor=\"white\">" CRLF
+"<center><h1>421 Misdirected Request</h1></center>" CRLF
+;
+
+
 static char ngx_http_error_494_page[] =
 "<html>" CRLF
 "<head><title>400 Request Header Or Cookie Too Large</title></head>"
@@ -251,26 +390,101 @@
 
 
 static char ngx_http_error_500_page[] =
-"<html>" CRLF
-"<head><title>500 Internal Server Error</title></head>" CRLF
-"<body bgcolor=\"white\">" CRLF
-"<center><h1>500 Internal Server Error</h1></center>" CRLF
+"<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">" CRLF
+"<html xmlns=\"http://www.w3.org/1999/xhtml\">" CRLF
+"<head>" CRLF
+"<meta http-equiv=\"Content-Type\" content=\"text/html; charset=iso-8859-1\"/>" CRLF
+"<title>500 - Internal server error.</title>" CRLF
+"<style type=\"text/css\">" CRLF
+"<!--" CRLF
+"body{margin:0;font-size:.7em;font-family:Verdana, Arial, Helvetica, sans-serif;background:#EEEEEE;}" CRLF
+"fieldset{padding:0 15px 10px 15px;} " CRLF
+"h1{font-size:2.4em;margin:0;color:#FFF;}" CRLF
+"h2{font-size:1.7em;margin:0;color:#CC0000;} " CRLF
+"h3{font-size:1.2em;margin:10px 0 0 0;color:#000000;} " CRLF
+"#header{width:96%;margin:0 0 0 0;padding:6px 2% 6px 2%;font-family:\"trebuchet MS\", Verdana, sans-serif;color:#FFF;" CRLF
+"background-color:#555555;}" CRLF
+"#content{margin:0 0 0 2%;position:relative;}" CRLF
+".content-container{background:#FFF;width:96%;margin-top:8px;padding:10px;position:relative;}" CRLF
+"-->" CRLF
+"</style>" CRLF
+"</head>" CRLF
+"<body>" CRLF
+"<div id=\"header\"><h1>Server Error</h1></div>" CRLF
+"<div id=\"content\">" CRLF
+" <div class=\"content-container\"><fieldset>" CRLF
+"  <h2>500 - Internal server error.</h2>" CRLF
+"  <h3>There is a problem with the resource you are looking for, and it cannot be displayed.</h3>" CRLF
+" </fieldset></div>" CRLF
+"</div>" CRLF
+"</body>" CRLF
+"</html>"
 ;
 
 
 static char ngx_http_error_501_page[] =
-"<html>" CRLF
-"<head><title>501 Not Implemented</title></head>" CRLF
-"<body bgcolor=\"white\">" CRLF
-"<center><h1>501 Not Implemented</h1></center>" CRLF
+"<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">" CRLF
+"<html xmlns=\"http://www.w3.org/1999/xhtml\">" CRLF
+"<head>" CRLF
+"<meta http-equiv=\"Content-Type\" content=\"text/html; charset=iso-8859-1\"/>" CRLF
+"<title>501 - Header values specify a method that is not implemented.</title>" CRLF
+"<style type=\"text/css\">" CRLF
+"<!--" CRLF
+"body{margin:0;font-size:.7em;font-family:Verdana, Arial, Helvetica, sans-serif;background:#EEEEEE;}" CRLF
+"fieldset{padding:0 15px 10px 15px;} " CRLF
+"h1{font-size:2.4em;margin:0;color:#FFF;}" CRLF
+"h2{font-size:1.7em;margin:0;color:#CC0000;} " CRLF
+"h3{font-size:1.2em;margin:10px 0 0 0;color:#000000;} " CRLF
+"#header{width:96%;margin:0 0 0 0;padding:6px 2% 6px 2%;font-family:\"trebuchet MS\", Verdana, sans-serif;color:#FFF;" CRLF
+"background-color:#555555;}" CRLF
+"#content{margin:0 0 0 2%;position:relative;}" CRLF
+".content-container{background:#FFF;width:96%;margin-top:8px;padding:10px;position:relative;}" CRLF
+"-->" CRLF
+"</style>" CRLF
+"</head>" CRLF
+"<body>" CRLF
+"<div id=\"header\"><h1>Server Error</h1></div>" CRLF
+"<div id=\"content\">" CRLF
+" <div class=\"content-container\"><fieldset>" CRLF
+"  <h2>501 - Header values specify a method that is not implemented.</h2>" CRLF
+"  <h3>The page you are looking for cannot be displayed because a header value in the request does not match certain configuration settings on the Web server. For example, a request header might specify a POST to a static file that cannot be posted to, or specify a Transfer-Encoding value that cannot make use of compression.</h3>" CRLF
+" </fieldset></div>" CRLF
+"</div>" CRLF
+"</body>" CRLF
+"</html>"
 ;
 
 
 static char ngx_http_error_502_page[] =
-"<html>" CRLF
-"<head><title>502 Bad Gateway</title></head>" CRLF
-"<body bgcolor=\"white\">" CRLF
-"<center><h1>502 Bad Gateway</h1></center>" CRLF
+"<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">" CRLF
+"<html xmlns=\"http://www.w3.org/1999/xhtml\">" CRLF
+"<head>" CRLF
+"<meta http-equiv=\"Content-Type\" content=\"text/html; charset=iso-8859-1\"/>" CRLF
+"<title>502 - Web server received an invalid response while acting as a gateway or proxy server.</title>" CRLF
+"<style type=\"text/css\">" CRLF
+"<!--" CRLF
+"body{margin:0;font-size:.7em;font-family:Verdana, Arial, Helvetica, sans-serif;background:#EEEEEE;}" CRLF
+"fieldset{padding:0 15px 10px 15px;} " CRLF
+"h1{font-size:2.4em;margin:0;color:#FFF;}" CRLF
+"h2{font-size:1.7em;margin:0;color:#CC0000;} " CRLF
+"h3{font-size:1.2em;margin:10px 0 0 0;color:#000000;} " CRLF
+"#header{width:96%;margin:0 0 0 0;padding:6px 2% 6px 2%;font-family:\"trebuchet MS\", Verdana, sans-serif;color:#FFF;" CRLF
+"background-color:#555555;}" CRLF
+"#content{margin:0 0 0 2%;position:relative;}" CRLF
+".content-container{background:#FFF;width:96%;margin-top:8px;padding:10px;position:relative;}" CRLF
+"-->" CRLF
+"</style>" CRLF
+"</head>" CRLF
+"<body>" CRLF
+"<div id=\"header\"><h1>Server Error</h1></div>" CRLF
+"<div id=\"content\">" CRLF
+" <div class=\"content-container\"><fieldset>" CRLF
+"  <h2>502 - Web server received an invalid response while acting as a gateway or proxy server.</h2>" CRLF
+"  <h3>There is a problem with the page you are looking for, and it cannot be displayed. When the Web server (while acting as a gateway or proxy) contacted the upstream content server, it received an invalid response from the content server.</h3>" CRLF
+" </fieldset></div>" CRLF
+"</div>" CRLF
+"</body>" CRLF
+"</html>"
 ;
 
 
@@ -334,8 +548,13 @@
     ngx_string(ngx_http_error_414_page),
     ngx_string(ngx_http_error_415_page),
     ngx_string(ngx_http_error_416_page),
+    ngx_null_string,                     /* 417 */
+    ngx_null_string,                     /* 418 */
+    ngx_null_string,                     /* 419 */
+    ngx_null_string,                     /* 420 */
+    ngx_string(ngx_http_error_421_page),
 
-#define NGX_HTTP_LAST_4XX  417
+#define NGX_HTTP_LAST_4XX  422
 #define NGX_HTTP_OFF_5XX   (NGX_HTTP_LAST_4XX - 400 + NGX_HTTP_OFF_4XX)
 
     ngx_string(ngx_http_error_494_page), /* 494, request header too large */
@@ -460,7 +679,6 @@
             case NGX_HTTPS_NO_CERT:
             case NGX_HTTP_REQUEST_HEADER_TOO_LARGE:
                 r->err_status = NGX_HTTP_BAD_REQUEST;
-                break;
         }
 
     } else {
@@ -616,14 +834,8 @@
     ngx_uint_t    msie_padding;
     ngx_chain_t   out[3];
 
-    if (clcf->server_tokens) {
-        len = sizeof(ngx_http_error_full_tail) - 1;
-        tail = ngx_http_error_full_tail;
-
-    } else {
-        len = sizeof(ngx_http_error_tail) - 1;
-        tail = ngx_http_error_tail;
-    }
+    len = sizeof(ngx_http_error_tail) - 1;
+    tail = ngx_http_error_tail;
 
     msie_padding = 0;
 
@@ -779,7 +991,7 @@
     b->last = ngx_cpymem(p, ngx_http_msie_refresh_tail,
                          sizeof(ngx_http_msie_refresh_tail) - 1);
 
-    b->last_buf = 1;
+    b->last_buf = (r == r->main) ? 1 : 0;
     b->last_in_chain = 1;
 
     out.buf = b;
diff -Naur ./nginx-1.10.0/src/http/v2/ngx_http_v2_filter_module.c ./nginx-iis-1.10.0/src/http/v2/ngx_http_v2_filter_module.c
--- ./nginx-1.10.0/src/http/v2/ngx_http_v2_filter_module.c	2016-04-26 21:31:20.000000000 +0800
+++ ./nginx-iis-1.10.0/src/http/v2/ngx_http_v2_filter_module.c	2017-03-15 18:08:22.144000882 +0800
@@ -143,7 +143,10 @@
 #endif
     u_char                     addr[NGX_SOCKADDR_STRLEN];
 
-    static const u_char nginx[5] = "\x84\xaa\x63\x55\xe7";
+#define NGINX_SERVER_NAME NGINX_VER
+
+    static size_t nginx_len = ngx_http_v2_literal_size(NGINX_SERVER_NAME);
+    static u_char nginx[ngx_http_v2_literal_size(NGINX_SERVER_NAME)];
 #if (NGX_HTTP_GZIP)
     static const u_char accept_encoding[12] =
         "\x8b\x84\x84\x2d\x69\x5b\x05\x44\x3c\x86\xaa\x6f";
@@ -230,7 +233,7 @@
     clcf = ngx_http_get_module_loc_conf(r, ngx_http_core_module);
 
     if (r->headers_out.server == NULL) {
-        len += 1 + (clcf->server_tokens ? nginx_ver_len : sizeof(nginx));
+        len += 1 + (clcf->server_tokens ? nginx_ver_len : sizeof(nginx_len));
     }
 
     if (r->headers_out.date == NULL) {
@@ -451,7 +454,7 @@
             pos = ngx_cpymem(pos, nginx_ver, nginx_ver_len);
 
         } else {
-            pos = ngx_cpymem(pos, nginx, sizeof(nginx));
+            pos = ngx_cpymem(pos, nginx, sizeof(nginx_len));
         }
     }
 
